numpy/core/src/multiarray/getset.c

```diff
@@ -492,58 +490,70 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
-            (PyArray_NDIM(self) == 0 ||
-             !PyArray_ISONESEGMENT(self) ||
-             PyDataType_HASSUBARRAY(newtype))) {
-        goto fail;
-    }
-
-    /* Deprecate not C contiguous and a dimension changes */
-    if (newtype->elsize != PyArray_DESCR(self)->elsize &&
-            !PyArray_IS_C_CONTIGUOUS(self)) {
-        /* 11/27/2015 1.11.0 */
-        if (DEPRECATE("Changing the shape of non-C contiguous array by\n"
-                      "descriptor assignment is deprecated. To maintain\n"
-                      "the Fortran contiguity of a multidimensional Fortran\n"
-                      "array, use 'a.T.view(...).T' instead") < 0) {
-            return -1;
+    /* Changing the size of the dtype results in a shape change */
+    if (newtype->elsize != PyArray_DESCR(self)->elsize) {
+        /* forbidden cases */
+        if (PyArray_NDIM(self) == 0) {
+            PyErr_SetString(PyExc_ValueError,
+                    "Changing the dtype of a 0d array is only supported "
+                    "if the itemsize is unchanged");
+            goto fail;
         }
-    }
-
-    if (PyArray_IS_C_CONTIGUOUS(self)) {
-        i = PyArray_NDIM(self) - 1;
-    }
-    else {
-        i = 0;
-    }
-    if (newtype->elsize < PyArray_DESCR(self)->elsize) {
-        /*
-         * if it is compatible increase the size of the
-         * dimension at end (or at the front for NPY_ARRAY_F_CONTIGUOUS)
-         */
-        if (PyArray_DESCR(self)->elsize % newtype->elsize != 0) {
+        else if (PyDataType_HASSUBARRAY(newtype)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "Changing the dtype to a subarray type is only supported "
+                    "if the total itemsize is unchanged");
             goto fail;
         }
-        newdim = PyArray_DESCR(self)->elsize / newtype->elsize;
-        PyArray_DIMS(self)[i] *= newdim;
-        PyArray_STRIDES(self)[i] = newtype->elsize;
-    }
-    else if (newtype->elsize > PyArray_DESCR(self)->elsize) {
-        /*
-         * Determine if last (or first if NPY_ARRAY_F_CONTIGUOUS) dimension
-         * is compatible
-         */
-        newdim = PyArray_DIMS(self)[i] * PyArray_DESCR(self)->elsize;
-        if ((newdim % newtype->elsize) != 0) {
+
+        /* determine which axis to resize */
+        if (PyArray_IS_C_CONTIGUOUS(self)) {
+            axis = PyArray_NDIM(self) - 1;
+        }
+        else if (PyArray_IS_F_CONTIGUOUS(self)) {
+            /* 2015-11-27 1.11.0 */
+            if (DEPRECATE(
+                        "Changing the shape of an F-contiguous array by "
+                        "descriptor assignment is deprecated. To maintain the "
+                        "Fortran contiguity of a multidimensional Fortran "
+                        "array, use 'a.T.view(...).T' instead") < 0) {
+                goto fail;
+            }
+            axis = 0;
+        }
+        else {
+            /* Don't mention the deprecated F-contiguous support */
```

Comment:
It's an else clause for `if(c-contig) elseif (f-contif) else`, so the obvious conclusion is `"must be c or f contig"`. But this would encourage the deprecated use.